Add support for non-SMT-COMP features in a new flag

mschlaipfer opened the pull request:
Some non-standard smtlib features are supported using a new flag -support-non-standard.
This first version allows to obfuscate/scramble some z3-specific smt lib files.

I think such an obfuscator is useful in general, and the scrambler already is such a tool. However, many solver-specific features are not supported right now. It seems to be easily extendable however, and non-standard features can be made explicit by checking for the new flag I have introduced.
The original code did not have to be touched.

Also added a flag to print the map of renamings.

mschlaipfer committed:
Add support for some z3 attributes

Supported using a new flag -support-non-standard.
This allows to obfuscate some z3-specific smt lib files.

Also added a flag to print the map of renamings.

mschlaipfer commented:
@aniemetz Hi, is there interest in merging this? This should be useful for applications such as https://bitbucket.org/viperproject/axiom-profiler/issues/14/enabled-sending-bug-reports-for in combination with more and more solver-based applications developed under NDAs in companies.

aniemetz reviewed (commented):
@mschlaipfer sorry, this somehow fell through the cracks.

`reset` is 'standard' in the sense that it is in the SMT-LIB standard. All of the others are Z3 extensions, right?

I'd suggest to have this more clearly documented, and maybe even distinguished with an extra option, i.e., one option for commands that are in the SMT-LIB standard but not included in SMT-COMP, and then one option per solver extensions, so in this case one extra option for the Z3 extensions.

Other than that I think it is indeed a good idea to include this in this tool. You're right, since we already have a scrambler, let's improve this one rather than writing a new one just because some people would find support for solver extensions useful. :+1:

aniemetz commented:
@mschlaipfer any updates on this?

mschlaipfer commented:
Sorry from me now :x, this has slipped off my to-do list. Thanks for your reply. I will add your suggestions next week and update the PR. I am not so familiar with different SMT solver's extensions, if it's okay I will stick to the Z3 ones I've added so far and move `reset` to 'standard'.

aniemetz commented:
> I will add your suggestions next week and update the PR.

:+1: 

> I am not so familiar with different SMT solver's extensions, if it's okay I will stick to the Z3 ones I've added so far and move `reset` to 'standard'.

Yes, I actually didn't mean to suggest that this PR should add support for other solver extensions, just that it would be good to have an option to explicitly enable Z3 extensions.

mschlaipfer committed:
Incorporated Aina's comments

added -support-non-smtcomp option, currently used for 'reset' and 'set-option'>
renamed -support-non-smtlib to -support-z3 and moved 'reset' out of it.

mschlaipfer commented:
I've updated the PR according to your comments, or at least my understanding of them. Please let me know if further changes are required.

aniemetz reviewed (commented):
Looks good!

One minor thing, can you clang-format your changes?
`git clang-format HEAD~2`
After that we're good to go.

mschlaipfer committed:
clang-format and manual formatting of parser.y

aniemetz commented:
@mschlaipfer thanks for your updates!
Do you maybe have two test cases we could check into the regression test suite?

mschlaipfer commented:
Are semantically nonsensical files okay? As long as they demonstrate the implemented features in the scrambler?

Should I integrate them into the `run_tests.sh` script? The script seems to be tailored to the SMTCOMP tracks, which those files would not be part of.

aniemetz commented:
Yes, semantically nonsensical files are fine.

By default, the test suite is run on all test files with all scrambler configurations used in the competition, so yes, we will actually need a different setup for these files.

I separated out the SMT-COMP tests, so you can go ahead and create a directory `test/extensions`, and enable them in `test/run_tests.sh` according to how the SMT-COMP tests are handled. Just make sure to use `${SCRIPT_DIR}/../scrambler` for the scrambler binary and pass all required options to enable the extensions.

mschlaipfer committed:
Remove -print-name-ids

mschlaipfer committed:
Merge branch 'master' into z3-specific

mschlaipfer committed:
Added extension tests, some fixes

fixed set-option
made skolemid a name node

mschlaipfer commented:
I have added the tests, it seems like you have (inadvertently?) removed all the others in the refactoring?

aniemetz commented:
Indeed. Readded them again.

aniemetz commented:
Hm, it seems that the test setup is misconfigured, Travis did not identify my commit, which should have been failing, as failing. I'll fix.

aniemetz commented:
Fixed

aniemetz commented on test/run_tests.sh:
```diff
@@ -5,7 +5,8 @@
 SCRIPT_DIR=$(dirname "$(readlink -f "$0")")
 TESTS_SMT_COMP_DIR="${SCRIPT_DIR}/smt-comp"
-EXPECT_SMT_COMP_DIR=${TESTS_SMT_COMP_DIR}/expect
+TESTS_NON_SMT_COMP_DIR="${SCRIPT_DIR}/extensions/non-smtcomp"
+TESTS_Z3_DIR="${SCRIPT_DIR}/extensions/z3"
```
You will now have to add checks if those two directories (and the corresponding `expect` directories) exist, similar to how it is done for the smt-comp directory on master.

mschlaipfer replied:
Sorry, missed that. Added now.

mschlaipfer committed:
Merge branch 'master' into z3-specific

mschlaipfer committed:
Add check for existence of new test dirs

aniemetz commented:
Thanks for all your work!

mschlaipfer commented:
Thanks for getting it in good shape :)
